File: 739-daily-temperatures/739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] res = new int[n];
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            if (stack.isEmpty()) {
                stack.add(i);
            } else {
                while (!stack.isEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                    res[stack.peek()] = i-stack.pop();
                }
                stack.add(i);
            }
        }
        return res;
    }
}

File: 739-daily-temperatures/739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> res(n);
        stack<int> stack;
        
        for (int i = 0; i < n; i++) {
            if (stack.empty()) {
                stack.push(i);
            } else {
                while (!stack.empty() && temperatures[stack.top()] < temperatures[i]) {
                    int j = stack.top(); 
                    res[j] = i-j;
                    stack.pop();
                }
                stack.push(i);
            }
        }
        return res;
    }
};